src/libYARP_OS/src/NetworkClock.cpp

```diff
@@ -9,131 +9,175 @@
 
-typedef std::list<std::pair<double, Semaphore*> > Waiters;
+typedef std::list<std::pair<double, Semaphore *> > Waiters;
 
 NetworkClock::NetworkClock()
-: pwaiters(YARP_NULLPTR), sec(0), nsec(0), t(0), closing(false){
+    : pwaiters(YARP_NULLPTR), sec(0), nsec(0), t(0), closing(false), clockName(""), initted(false)
+{
     pwaiters = new Waiters();
     yAssert(pwaiters!=YARP_NULLPTR);
 }
 
-NetworkClock::~NetworkClock() {
+NetworkClock::~NetworkClock()
+{
     listMutex.lock();
     closing = true;
     port.interrupt();
 
-    Waiters* waiters = static_cast<Waiters*>(pwaiters);
-    if (waiters) {
+    Waiters *waiters = static_cast<Waiters *>(pwaiters);
+
+    if (waiters)
+    {
         Waiters::iterator waiter_i;
         waiter_i = waiters->begin();
+
         while (waiter_i != waiters->end())
         {
             Semaphore *waiterSemaphore = waiter_i->second;
             waiter_i = waiters->erase(waiter_i);
+
             if (waiterSemaphore)
                 waiterSemaphore->post();
         }
-        listMutex.unlock();
 
         delete waiters;
         pwaiters = YARP_NULLPTR;
     }
-    else {
-        listMutex.unlock();
-    }
+
+    listMutex.unlock();
+    yarp::os::ContactStyle style;
+    style.persistent = true;
+    NetworkBase::disconnect(clockName, port.getName(), style);
 }
 
 
-bool NetworkClock::open(const ConstString& name) {
+bool NetworkClock::open(const ConstString &clockSource_portName)
+{
     port.setReadOnly();
     port.setReader(*this);
-    NestedContact nc(name);
-    if (nc.getNestedName()=="") {
-        Contact src = NetworkBase::queryName(name);
-        if (src.isValid()) {
-            bool ok = port.open("");
-            if (!ok) return false;
-            return NetworkBase::connect(name,port.getName());
-        } else {
-            fprintf(stderr,"Cannot find time port \"%s\"; for a time topic specify \"%s@\"\n", name.c_str(), name.c_str());
-            return false;
-        }
+    NestedContact nc(clockSource_portName);
+    clockName = clockSource_portName;
+    yarp::os::ContactStyle style;
+    style.persistent = true;
+//     style.persistenceType = ContactStyle::END_WITH_TO_PORT;
+
+
+    // this should be an anonymous port to not pollute the yarp name list  ( or use the $PID/clock:in ?)
+    ConstString myName = "...";
+
+    //    #include <yarp/conf/system.h>
+    //    pid = ACE_OS::getpid();
+
+    // if receiving port cannot be opened, return false.
+    bool ret = port.open(myName);
+    if (!ret)
+        return false;
+
+    if (nc.getNestedName()=="")
+    {
+        Contact src = NetworkBase::queryName(clockSource_portName);
+
+
+        ret = NetworkBase::connect(clockSource_portName, port.getName(), style);
+
+        if(!src.isValid())
+            fprintf(stderr,"Cannot find time port \"%s\"; for a time topic specify \"%s@\"\n", clockSource_portName.c_str(), clockSource_portName.c_str());
     }
-    return port.open(name);
+
+    delay(0.01);
```

Comment:
This PR is a collection of workaround, not the final implementation, so those are some dirty fix **_but_** with a reason. Those will be replaced with something better in the next version but right now there is no way of doing it better.
The reason here is that when the function returns from the openNetworkClock, the clock _needs to be initialize_ with a valid time value.
The delay basically tries to synch the local clock receiver with the clock source in a  `wait for the first message to arrive before returning from open` fashion. 

Scenario: I instantiate the network clock and then ask for a `time::now()`. It may happen that the ask for `time::now` arrives before the first yarp message delivering the time. Which time shall the time::now return? If it returns zero, then the first yarp time message contains the value `1000seconds`, the clock will jump forward of 1000 sec which is bad.
This way the instantiation of the clock will not leave the open until I have an actual valid value.
Of course, this also means that if no clock producer is active, the program will get stuck in the network::init (which may be annoying) but it is a problem of the system as a whole and not of the particular program under execution.
Furthermore it is easier to detect this problem at the startup, then starting the program with a network clock which is not receiving any data.

Again, it is not the cleanest way and I'll figure out something better for the final version, but it works.